fix the race described in #3174

himanshug opened the pull request:
trying the fix to #3174 

this patch makes worker selection from strategy and its insert into workersWithUnacknowledgedTask atomic by putting both into a synchronized block so that multiple pending task runner threads never get same worker.

fjy commented:
@himanshug is there any way to add a test to verify things?

himanshug commented:
@fjy this one is pretty hard to unit test

nishantmonu51 commented:
this seems to be introducing a regression for fill capacity worker selection strategy. 
with single assignment thread + fill capacity worker selection strategy, all the tasks get assigned to same worker until its capacity gets filled and only when a worker is filled to its capacity tasks are assigned to next worker. 

With above patch, with multiple assignment threads and fill capacity worker select strategy, it is possible that each assignment thread will assign tasks to a separate worker even though none of them is filled to their capacity. 

This may not affect most of the setups practically, but it theoretically alters the contract for worker selection strategy.

drcrallen commented:
@himanshug Doesn't this proposed method of solving this effectively eliminate the benefit of multi-threaded executor for submitting tasks?

Somehow we need to preserve the ability to assign tasks to "behaving" workers in a world where some of the workers are misbehaving and taking along time to acknowledge tasks.

If we take a purely parallel approach per worker, such no 2 launches can be in flight at the same time for the same worker, then that could violate the fill capacity contract, which in general would be OK. But there are still corner cases for things like a single worker with ample capacity to launch tasks, but is misbehaving, what is the proper way to handle this? It would appear as if the cluster has capacity, but in fact does not.

himanshug commented:
@drcrallen  i don't think so, synchronization is only being introduced for 2 operations..
  - select a worker using selectorStrategy
  - add selected worker to workersWithUnacknowledgedTask
this is done to ensure that same worker is not given to more than one thread.
announceTask step still happens in parallel. So, if there was a worker taking long time to ack the task then other threads will continue making progress with other workers. That was the original intention behind #2521

@nishantmonu51  That is true, we can document this caveat in the doc, that when user decides to use more than one thread then fill capacity might not work as expected.

nishantmonu51 commented:
@himanshug documenting the caveat seems fine. 
Also, How do you feel about just doing task announcement in a multi threaded exec instead of making complete runPendingTasks method multithreaded ?
other tasks for selecting a task to assign and mapping it to worker can still be done in a single thread (We are anyways adding sync blocks for it).

I believe that might make the code simpler and easy to understand.
We may even be able to abstract out the logic for task announcement in a separate MultithreadedTaskAnnouncer class.

drcrallen commented:
@himanshug gotcha, my visual inspection was putting the `}` in the wrong place

himanshug commented:
@nishantmonu51 that sounds interesting... i will try that.

fjy commented:
@himanshug any update on this PR?

fjy commented:
ping @himanshug

himanshug commented:
@fjy thanks for reminders :)

tried to implement @nishantmonu51 's suggestion and got into few issues ... we discussed in the dev-syncup and will further update with the conclusion.

nishantmonu51 commented:
:+1: after documenting the caveat that fill worker with capacity might not work as expected with multi threaded task assignment.

himanshug commented:
updated the doc. also changed default pendingTasksRunnerNumThreads = 1 ... we will internally keep it 3 and continue to monitor if any issues are faced.

fjy commented:
👍
